### ISSUE TITLE
Unable to use a custom STI OAuthApplication class

### ISSUE THREAD

**Issue by hmcfletch** [doorkeeper-gem/doorkeeper#1513]
### Steps to reproduce
Here is the basic setup:

I added a `type` column to the `oauth_applications` database table.

I'm using something similar to the following class setup:
```ruby
class MyCustomOauthApplicationClass  < ApplicationRecord
  include ::Doorkeeper::Orm::ActiveRecord::Mixins::Application
end

class MySpecialCustomOauthApplicationClass < MyCustomOauthApplicationClass
end
```

I have a bit of custom logic for the `AuthorizationsController` to integrate with our SPA. This shouldn't be affecting things, since I was able to reproduce the issue in the Rails console. Just adding for some completeness.
```ruby
class OauthIdpAuthorizationsController < Doorkeeper::AuthorizationsController
  def new
    if pre_auth.authorizable?
      redirect_to spa_app_path(spa_route: 'oauth/spa_authorize', **pre_auth_params)
    else
      render_error
    end
  end
end
```

### Expected behavior
`pre_auth` in `Doorkeeper::AuthorizationsController` should be able to find and initialize the correct model instance. 

### Actual behavior
I get the following error:
`ActiveRecord::SubclassNotFound: Invalid single-table inheritance type: MySpecialCustomOauthApplicationClass is not a subclass of MyCustomOauthApplicationClass`

### What I have been able to figure out
I went on a deep dive into ActiveRecord's internals to hunt down how STI was handled. That lead me to `ActiveSupport::DescendantsTracker` which, as the name implies, tracks all of the subclasses, recursively, of a class. 

It looks like the object that is held in `Doorkeeper.config.application_model` is different than calling `MyCustomOauthApplicationClass`.

They actually return different `object_id`s which is odd. This is a scrubbed version of a console session:
```ruby
[1] > MyCustomOauthApplicationClass.object_id
=> 97020
[2] > Doorkeeper.config.application_model.name
=> "MyCustomOauthApplicationClass"
[3] > Doorkeeper.config.application_model.object_id
=> 90500
[4] > uid = 'asfsafsdf' # uid of an STI class instance
=> "asfsafsdf"
[5] > MyCustomOauthApplicationClass.find_by(uid: uid)
=> #<MySpecialCustomOauthApplicationClass id: 2, ...>
[6] > Doorkeeper.config.application_model.find_by(uid: uid)
=> ActiveRecord::SubclassNotFound: Invalid single-table inheritance type: MySpecialCustomOauthApplicationClass is not a subclass of MyCustomOauthApplicationClass
[7] > ActiveSupport::DescendantsTracker.descendants(MyCustomOauthApplicationClass)
=> [MySpecialCustomOauthApplicationClass(id: integer, type: string, ...)]
[8] > ActiveSupport::DescendantsTracker.descendants(Doorkeeper.config.application_model)
=> []
```

`ActiveSupport::DescendantsTracker` is a critical part of determining the correctness of STI classes. It looks like two different versions of `MyCustomOauthApplicationClass` are getting loaded and the one associated with `Doorkeeper.config.application_model` isn't getting setup correctly.

I'm not an expert on class, constant, or object loading, but I wonder if the fact that `Doorkeeper.config.application_model` `object_id` is lower means that it was loaded first but not in a way that ActiveRecord can find it and use it.

Possibly related issues:
- #1492
- #1489

### System configuration
**Doorkeeper initializer**:

```ruby
# config/initializers/doorkeeper.rb
Doorkeeper.configure do
  orm :active_record
  application_class 'MyCustomOauthApplicationClass'

  api_only
end

# routes.rb
use_doorkeeper do
  controllers authorizations: 'oauth_idp_authorizations'
  skip_controllers :applications, :token_info, :authorized_applications
end
```

**Ruby version**: `2.7.3`

**Gemfile.lock**:

<details>
  <summary>Gemfile.lock content</summary>
  
```
activerecord (6.1.3.2)
doorkeeper (5.5.1)
```

Let me know if there are any gem versions you need beyond these.
</details>

**Comment by hmcfletch** [comment]
I am also seeing the file that defines `MyCustomOauthApplicationClass` loaded twice as well.

**Comment by hmcfletch** [comment]
Digging deeper it looks like a call to `Doorkeeper.config.application_model` triggers the model load via `ActiveSupport::Dependencies` and calling the model directly triggers a `require` via `zeitwerk`. It looks like these two have separate class caches which don't interact.

**Comment by hmcfletch** [comment]
Well this was a doozy...

Setting this in one of my initializers fixed the issue.

```ruby
ActiveSupport::Dependencies.mechanism = :require
```

**Comment by nbulaj** [comment]
Thanks @hmcfletch for the issue and the investigation. Yeap, could be that Rails autoloading mechanism do some weird issues. I have to think about better ways of lazy-laoading for the models

**Comment by fxn** [comment]
This smells like a stale class object is alive.

Which version of Rails are you using? Seems to be using the classic autoloader, right? Is the application autoloading anything during initialization?

**Comment by hmcfletch** [comment]
Using Rails 6. Not doing anything interesting in my initializers related to loading. If I remember correctly setting:
```ruby
ActiveSupport::Dependencies.mechanism = :require
```
Fixed the STI issues for the initial load in dev, but if you reload the app STI classes break.

**Comment by fxn** [comment]
@hmcfletch What does

```
bin/rails runner 'p Rails.autoloaders.zeitwerk_enabled?'
```

print?

Can you check `log/development.log` and see if there are warnings about reloadable constants being loaded during initialization?

**Comment by fxn** [comment]
OK, I have a theory.

Your application is autoloading during initialization.

While 

```ruby
application_class 'MyCustomOauthApplicationClass'
```

correctly receives a string instead of a class object, a posteriori it would agree with your report that [`application_model`](https://github.com/doorkeeper-gem/doorkeeper/blob/a56b1a5379e97fed5cf838f834702724e070db38/lib/doorkeeper/config.rb#L464) is evaluated (by Doorkeeper itself internally?) before initialization finishes.

In order to prevent these types of problems, Rails 6 issues warnings about autoloading reloadable constants, and in Rails 7 that is an error condition and the application would not boot.

You can verify this by putting

```
pp caller_locations
```

at the top of `my_custom_oauth_application_class.rb`.

The workaround

```ruby
ActiveSupport::Dependencies.mechanism = :require
```

works because you load the autoloaded classes with `require`, and Zeitwerk ignores those files. In consequence, those classes are not reloadable any more. Changes in them would have no effect.

But we need to find the root of the problem, that workaround may unblock you, but is not the true fix. To suggest a fix I need first to understand why `application_model` is evaluated so early. That defeats the point of receiving a string as `application_class`.

(`ActiveSupport::Dependencies.mechanism=` does not even exist in Rails 7, that was a setter of the `classic` autoloader, which is the one autoloading during initialization due to a compromise taken in the integration during the transition.)

**Comment by hmcfletch** [comment]
Dug in a little before I saw your follow up and I am thinking the same things about the autoloading situation. 

`zeitwerk_enabled?` is `true`, but I get a number of things showing up in `DEPRECATION WARNING: Initialization autoloaded the constants`. I'll take a look at fixing these issues and get back to you. This feels like the root of the issue.

Thanks for taking the time to think on this and provide some education on the topic. I have dug into a bunch of the loading code on my own, but your above comment definitely gave me a bit of a lightbulb moment. Thanks!

**Comment by nbulaj** [comment]
@fxn Doorkeeper configuration process is [run](https://github.com/doorkeeper-gem/doorkeeper/blob/main/lib/doorkeeper/config.rb#L24) in the theinitializer (`config/initializers/doorkeeper.rb`) and it setup ORM dependent files there. During configuration it [constantizes ORM ](https://github.com/doorkeeper-gem/doorkeeper/blob/3c74834f9c190fccaa958be394710c08e5d33b9c/lib/doorkeeper/config.rb#L40-L41) and then initializes the models where [it requires all required ORM files](https://github.com/doorkeeper-gem/doorkeeper/blob/3c74834f9c190fccaa958be394710c08e5d33b9c/lib/doorkeeper/orm/active_record.rb#L16-L21).

I think the culprit is there, but don't sure where exactly. Also some time ago I thought about moving this to `to_prepare`, but don't remember why I didn't do it :thinking: And what was the reasoning.

**Comment by fxn** [comment]
Didn't have time to do a full walkthrough, but let me share some first thoughts.

Let's recap reloading for a moment. When Rails reloads `MyCustomOauthApplicationClass`, it deletes the constant `MyCustomOauthApplicationClass` from `Object`. If you evaluate `MyCustomOauthApplicationClass` later, the corresponding file is loaded again. As a side-effect `MyCustomOauthApplicationClass` is defined and **stores a new class object**.

In Rails 6, if you autoload during initialization with eager loading disabled, those autoloaded constants are unloaded (the warnings says so). It does so precisely to make latent issues surface. Hence, evaluating `MyCustomOauthApplicationClass` again after boot yields a new class object. The need to load the STI at runtime may be triggering that.

OK, let's consider

```ruby
def application_model
  @application_model ||= application_class.constantize
end
```

Since the application model is memoized in that ivar, reloading has no effect within Doorkeeper. Nothing is refreshing that object (I am guessing, please let me know if I overlooked something). The original class object is cached there, it's a stale object.

In a library that supports user-level application classes, the key to work well with reloading is:

1. Configuration should ask for class names, not class objects. Doorkeeper already does this ✅.
2. `constantize` on that string is executed after boot. If the value is cached, then you reset in a `Rails.application.config.to_prepare` block.

Would it fit the design of the library and its initialization to have a `Rails.application.config.to_prepare` block in which `application_model` and its setup run again? Maybe that would suffice.

Alternatively, if the design requires that the application model be initialized early and cached forever, then an alternative would be to document it and explain how to avoid reloading those classes.

**Comment by Flixt** [comment]
I'm not sure about all the autoloader/zeitwerk related stuff but for me the error occurs  at

https://github.com/doorkeeper-gem/doorkeeper/blob/v5.5.4/lib/doorkeeper/orm/active_record.rb#L35

when I use `application_class 'OAuthApplication'` together with following line in the `config/initializers/doorkeeper.rb`:

```
enable_application_owner confirmation: false
```

It happens always when I boot my Rails application (server, console).

Putting a `require oauth_application'  before the line `Doorkeeper.config.application_model.include(Doorkeeper::Models::Ownership)`
removes the error `uninitialized constant OAuthApplication (NameError)`

Pretty sure the require should not be necessary, just wanted to point out were exactly the early evaluation of the user-level application class happens and leads to the error for me.

**Comment by nbulaj** [comment]
@hmcfletch @Flixt could you please try with my old experiments around Ruby `autoload` and Rails `to_prepare` ?

```ruby
gem 'doorkeeper', git: "https://github.com/doorkeeper-gem/doorkeeper.git", branch: "fix-models-autoload"
```

**Comment by fxn** [comment]
Hey! On my phone!

I think it would help to have a really minimal app using SQLite that allow
us to experiment and share observations? Could you guys prepare one?

Let me clarify also, just in case, that autoloading doesn't happen in
doorkeeper's stuff under `lib`. That's loaded normally, and it is not
reloaded.

El dt, 7 des 2021 a les 15:21 Nikita Bulai ***@***.***> va
escriure:

**Comment by nbulaj** [comment]
The branch with autoloading excludes calls to models while initializing and calls to_prepare for some specific logic, that';s why I assume it should be OK. But... :smile: we'll see

**Comment by hmcfletch** [comment]
I'll be trying this out today or tomorrow 👍

**Comment by hmcfletch** [comment]
Getting close! 

I was getting `NameError: uninitialized constant Doorkeeper::Orm::ActiveRecord::Mixins`, so I had to add
```ruby
require 'doorkeeper/orm/active_record/mixins/application'
```
to my AR STI application class. I suspect I'd need to do something similar to my other custom classes as well. After that, everything seemed to be good. Even after a `reload!`

**Comment by sigra** [comment]
@nbulaj your branch `fix-models-autoload` works good and solves issue for our company. Is the any chance to have this merged into the main release?

**Comment by nbulaj** [comment]
Hi @sigra .

[This branch](https://github.com/doorkeeper-gem/doorkeeper/pull/1551) requires fixes for tests and then could me merged. Also I'm worried with such changes so it will be releases at least as a pre version so developers could check how it works on their codebase.

**Comment by nbulaj** [comment]
@fxn what do you generally think about memoizing constantized values? Does it makes sense for some performance reasons (or any other) or it's better to do call it (`.constantize`) every time method is called?

**Comment by fxn** [comment]
@nbulaj nowadays `constantize` is basically `const_get`. However, if the value is cached in a place that gets reloaded, I believe I'd personally prefer to cache it. Even if the method is cheap, I believe I'd have a sense of doing unnecessary method calls nonetheless.

**Comment by nbulaj** [comment]
@hmcfletch @sigra could you please check with the latest changes from https://github.com/doorkeeper-gem/doorkeeper/pull/1551 ? I fixed the specs and most of the issues. 

Note that it;s  not merged  to `main` yet
